File: src/ios/CordovaCommerceWeb.m

/* Copyright © 2020 Mastercard. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 =============================================================================*/


#import "CordovaCommerceWeb.h"

static NSString* toBase64(NSData* data) {
    SEL s1 = NSSelectorFromString(@"cdv_base64EncodedString");
    SEL s2 = NSSelectorFromString(@"base64EncodedString");
    SEL s3 = NSSelectorFromString(@"base64EncodedStringWithOptions:");

    if ([data respondsToSelector:s1]) {
        NSString* (*func)(id, SEL) = (void *)[data methodForSelector:s1];
        return func(data, s1);
    } else if ([data respondsToSelector:s2]) {
        NSString* (*func)(id, SEL) = (void *)[data methodForSelector:s2];
        return func(data, s2);
    } else if ([data respondsToSelector:s3]) {
        NSString* (*func)(id, SEL, NSUInteger) = (void *)[data methodForSelector:s3];
        return func(data, s3, 0);
    } else {
        return nil;
    }
}

@implementation CordovaCommerceWeb

/// Plugin initialization
-(void)pluginInitialize{
    NSLog(@"Mastercard CordovaCommerceWeb plugin initialized");
}

-(void)initializeSdk:(CDVInvokedUrlCommand *)command{
    self.callbackId = command.callbackId;
    CDVPluginResult* pluginResult = nil;
    NSDictionary* configDic = [command.arguments objectAtIndex:0];
    if (configDic != nil && configDic.count > 0) {
        NSLocale *locale = [[NSLocale alloc] initWithLocaleIdentifier:configDic[@"locale"]];
        NSString *checkoutId = configDic[@"checkoutId"];
        NSString *checkoutUrl = configDic[@"checkoutUrl"];
        NSString *callbackScheme = configDic[@"callbackScheme"];
        NSSet *allowedCardTypes = [NSSet setWithArray:configDic[@"allowedCardTypes"]];
        MCSConfiguration *config = [[MCSConfiguration alloc] initWithLocale:locale
                                                                 checkoutId:checkoutId
                                                                checkoutUrl:checkoutUrl
                                                             callbackScheme:callbackScheme
                                                           allowedCardTypes:allowedCardTypes
                                                   presentingViewController:nil];

        MCSCommerceWeb *commerceWeb = [MCSCommerceWeb sharedManager];
        [commerceWeb initWithConfiguration:config];
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:configDic.description];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
    }
    
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

-(void)checkout:(CDVInvokedUrlCommand *)command{
    self.callbackId = command.callbackId;
    NSDictionary* checkoutDic= [command.arguments objectAtIndex:0];
    MCSCheckoutRequest *checkoutRequest = [[MCSCheckoutRequest alloc] init];
    double amount = [[checkoutDic valueForKey:@"amount"] doubleValue];
    checkoutRequest.amount = [[NSDecimalNumber alloc] initWithDouble:amount];
    checkoutRequest.currency = checkoutDic[@"currency"];
    checkoutRequest.cartId = checkoutDic[@"cartId"];
    _Bool supress = checkoutDic[@"suppressShippingAddress"];
    checkoutRequest.suppressShippingAddress = !supress;
    
    checkoutRequest.callbackUrl = checkoutDic[@"calbackUrl"];
    checkoutRequest.unpredictableNumber = @"12345678";
    
    MCSCryptoOptions *cryptoOptionMaster = [[MCSCryptoOptions alloc] init];
    cryptoOptionMaster.cardType = MCSCardTypeMaster;
    cryptoOptionMaster.format = [NSSet setWithArray:checkoutDic[@"cryptoOptions"]];
    checkoutRequest.cryptoOptions = @[cryptoOptionMaster];
    MCSCommerceWeb *commerceWeb = [MCSCommerceWeb sharedManager];
    [commerceWeb setDelegate:self];
    [commerceWeb checkoutWithRequest:checkoutRequest];
}

-(void)getCheckoutButton:(CDVInvokedUrlCommand *)command{
    MCSCommerceWeb *commerceWeb = [MCSCommerceWeb sharedManager];
    MCSCheckoutButton *button = [commerceWeb checkoutButtonWithDelegate:self];
    //convert the image to NSData first
    NSData *nsdata = UIImagePNGRepresentation(button.imageView.image);
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:toBase64(nsdata)];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)checkoutRequest:(MCSCheckoutRequest *)request didCompleteWithStatus:(MCSCheckoutStatus)status forTransaction:(NSString *)transactionId{
    CDVPluginResult* pluginResult = nil;
    switch (status) {
        case MCSCheckoutStatusSuccess:
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:transactionId];
            break;
        case MCSCheckoutStatusCanceled:
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Can't complete Checkout"];
            break;
        default:
             pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Can't complete Checkout"];
            break;
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:self.callbackId];
}

- (void)checkoutRequestForTransaction:(nonnull void (^)(MCSCheckoutRequest * _Nonnull))handler {}

@end

File: src/android/CordovaCommerceWeb.java
package org.apache.cordova.plugin.mastercard;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.util.Base64;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.mastercard.commerce.CardType;
import com.mastercard.commerce.CheckoutButton;
import com.mastercard.commerce.CheckoutCallback;
import com.mastercard.commerce.CheckoutRequest;
import com.mastercard.commerce.CommerceConfig;
import com.mastercard.commerce.CommerceWebSdk;
import com.mastercard.commerce.CryptoOptions;
import com.mastercard.commerce.Mastercard;

import org.apache.cordova.*;
import org.json.JSONArray;
import org.json.JSONException;

import java.io.ByteArrayOutputStream;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

/* Copyright © 2020 Mastercard. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 =============================================================================*/


/**
 * This class is  used to call CommerceWeb SDK functionality from JavaScript.
 */
public class CordovaCommerceWeb extends CordovaPlugin {

  private static final String TAG = CordovaCommerceWeb.class.getSimpleName();
  private static final String ACTION_INITIALIZE_SDK = "initializeSdk";
  private static final String ACTION_CHECKOUT = "checkout";
  private static final String ACTION_GET_CHECKOUT_BUTTON = "getCheckoutButton";
  private static final String CARD_TYPE_MASTER = "master";
  private static final String CARD_TYPE_VISA = "visa";
  private static final String CARD_TYPE_AMEX = "amex";
  private static final String CARD_TYPE_JCB = "jcb";
  private static final String CARD_TYPE_DISCOVER = "discover";
  private static final String CARD_TYPE_MAESTRO = "maestro";
  private static final String CARD_TYPE_DINER = "diner";
  private static final String CRYPTO_OPTIONS_TYPE_ICC = "ICC";
  private static final String CRYPTO_OPTIONS_TYPE_UCAF = "UCAF";
  private static final int IMAGE_QUALITY_HIGH = 100;
  private static final String TRANSACTION_ID = "TransactionId";

  private CommerceWebSdk commerceWebSdk;
  private Gson gson = new GsonBuilder().create();
  private CallbackContext callbackContext = null;

  @Override
  public void initialize(CordovaInterface cordova, CordovaWebView webView) {
      super.initialize(cordova, webView);
      Log.d(TAG, "Mastercard CordovaCommerceWeb plugin initialized");
  }

  @Override
  public void onNewIntent(Intent intent) {
    //handle call back transaction ID back to JS
    if(intent.getExtras()!=null){
      PluginResult result = null;
      if(intent.getExtras().getString(TRANSACTION_ID) != null){
        String transcationID = intent.getExtras().getString(TRANSACTION_ID);
        result = new PluginResult(PluginResult.Status.OK,transcationID);
      }else{
        result = new PluginResult(PluginResult.Status.ERROR, "Expected transaction ID non-empty.");
      }
      this.callbackContext.sendPluginResult(result);
    }
  }

  @Override
  public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
    this.callbackContext = callbackContext;
    if (action.equalsIgnoreCase(ACTION_INITIALIZE_SDK)) {
        this.initializeSdk(args.getString(0),callbackContext);
      return true;
    } else if (action.equalsIgnoreCase(ACTION_CHECKOUT)) {
        this.checkout(args.getString(0),callbackContext);
      return true;
    } else if (action.equalsIgnoreCase(ACTION_GET_CHECKOUT_BUTTON)) {
        this.getCheckoutButton(args.getString(0), callbackContext);
      return true;
    }

    return false;
  }

  private void initializeSdk(String configurationString,final CallbackContext callbackContext) {
    PluginResult result = null;
    if (configurationString != null && configurationString.length() > 0) {
        ConfigurationModel configModel = gson.fromJson(configurationString, ConfigurationModel.class);
        CommerceConfig config =
            new CommerceConfig(configModel.getLocale(), configModel.getCheckoutId(), configModel.getCheckoutUrl(), getAllowedCardTypes(configModel));

        commerceWebSdk = CommerceWebSdk.getInstance();
        commerceWebSdk.initialize(cordova.getActivity().getApplicationContext(), config);
        result = new PluginResult(PluginResult.Status.OK, configurationString);
        this.callbackContext.sendPluginResult(result);
    } else {
        result = new PluginResult(PluginResult.Status.ERROR, "Expected initializeSdk non-empty string argument.");
        this.callbackContext.sendPluginResult(result);
    }
  }

  private void checkout(String checkoutRequestString,final CallbackContext callbackContext) {
    PluginResult result = null;
    if (checkoutRequestString != null && checkoutRequestString.length() > 0) {
        CheckoutRequest request = buildCheckoutRequest(checkoutRequestString);
        commerceWebSdk.checkout(request);
    } else {
      result = new PluginResult(PluginResult.Status.ERROR, "Expected checkout non-empty string argument.");
      this.callbackContext.sendPluginResult(result);
    }
  }

  private void getCheckoutButton(String checkoutRequestString, final CallbackContext callbackContext) {
    if (checkoutRequestString != null && checkoutRequestString.length() > 0) {
        CheckoutButton checkoutButton = commerceWebSdk.getCheckoutButton(getCheckoutCallback(checkoutRequestString));
        String checkoutButtonImage = getBase64ImageString(checkoutButton.getDrawable(), IMAGE_QUALITY_HIGH);
        this.callbackContext.success(checkoutButtonImage);
    } else {
        this.callbackContext.error("Expected getCheckoutButton non-empty string argument.");
    }
  }


  private CheckoutCallback getCheckoutCallback(final String checkoutRequestString) {

    return listener -> {
      CheckoutRequest request = buildCheckoutRequest(checkoutRequestString);
      listener.setRequest(request);
    };
  }

  private CheckoutRequest buildCheckoutRequest(String checkoutRequestString) {
    CheckoutRequestModel checkoutRequestModel = gson.fromJson(checkoutRequestString, CheckoutRequestModel.class);

    return new CheckoutRequest.Builder().amount(checkoutRequestModel.getAmount())
            .cartId(checkoutRequestModel.getCartId())
            .currency(checkoutRequestModel.getCurrency())
            .callbackUrl(checkoutRequestModel.getCallbackUrl())
            .cryptoOptions(getCryptoOptions(checkoutRequestModel))
            .cvc2Support(checkoutRequestModel.isCvc2Support())
            .shippingLocationProfile(checkoutRequestModel.getShippingLocationProfile())
            .suppress3ds(checkoutRequestModel.isSuppress3Ds())
            .suppressShippingAddress(checkoutRequestModel.isSuppressShippingAddress())
            .unpredictableNumber(checkoutRequestModel.getUnpredictableNumber())
            .validityPeriodMinutes(checkoutRequestModel.getValidityPeriodMinutes())
            .build();
  }

  private Set<CardType> getAllowedCardTypes(ConfigurationModel configurationModel) {
    Set<CardType> allowedCardTypeSet = new HashSet<>();

    for (String cardType : configurationModel.getAllowedCardTypes()) {
      if (cardType.equalsIgnoreCase(CARD_TYPE_MASTER)) {
        allowedCardTypeSet.add(CardType.MASTER);
      } else if (cardType.equalsIgnoreCase(CARD_TYPE_VISA)) {
        allowedCardTypeSet.add(CardType.VISA);
      } else if (cardType.equalsIgnoreCase(CARD_TYPE_AMEX)) {
        allowedCardTypeSet.add(CardType.AMEX);
      } else if (cardType.equalsIgnoreCase(CARD_TYPE_JCB)) {
        allowedCardTypeSet.add(CardType.JCB);
      } else if (cardType.equalsIgnoreCase(CARD_TYPE_DISCOVER)) {
        allowedCardTypeSet.add(CardType.DISCOVER);
      } else if (cardType.equalsIgnoreCase(CARD_TYPE_MAESTRO)) {
        allowedCardTypeSet.add(CardType.MAESTRO);
      } else if (cardType.equalsIgnoreCase(CARD_TYPE_DINER)) {
        allowedCardTypeSet.add(CardType.DINERS);
      }
    }

    return allowedCardTypeSet;
  }

  private Set<CryptoOptions> getCryptoOptions(CheckoutRequestModel checkoutRequestModel) {
    Set<Mastercard.MastercardFormat> mastercardFormatSet = new HashSet<>();

    for (String cryptoOptionsType : checkoutRequestModel.getCryptoOptions()) {
      if (cryptoOptionsType.equalsIgnoreCase(CRYPTO_OPTIONS_TYPE_ICC)) {
        mastercardFormatSet.add(Mastercard.MastercardFormat.ICC);
      } else if (cryptoOptionsType.equalsIgnoreCase(CRYPTO_OPTIONS_TYPE_UCAF)) {
        mastercardFormatSet.add(Mastercard.MastercardFormat.UCAF);
      }
    }

    CryptoOptions mastercard = new Mastercard(mastercardFormatSet);
    Set<CryptoOptions> cryptoOptionsSet = new HashSet<>();
    cryptoOptionsSet.add(mastercard);

    return cryptoOptionsSet;
  }

  private String getBase64ImageString(Drawable drawable, int imageQuality) {
    Bitmap anImage = ((BitmapDrawable) drawable).getBitmap();

    return encodeToBase64(anImage, Bitmap.CompressFormat.PNG, imageQuality);
  }

  private String encodeToBase64(Bitmap image, Bitmap.CompressFormat compressFormat, int quality) {
    ByteArrayOutputStream byteArrayOS = new ByteArrayOutputStream();
    image.compress(compressFormat, quality, byteArrayOS);

    return Base64.encodeToString(byteArrayOS.toByteArray(), Base64.DEFAULT);
  }

}